### PR TITLE
Feature/user sequence number

### PR DESCRIPTION
**Pull request by hammad-Ikhlaq-7vals**
Added Sequence Id

### CODE REVIEW

File: app/controllers/companies_controller.rb [hunk 1]
```diff
@@ -4,6 +4,10 @@ def company
     @company = current_company
     @projects = Project.all
     @issues = Issue.all
-    flash.now[:notice] = t('companies.welcome_message', company_name: @company.name)
+    @users = User.all
+    flash.now[:notif] = "Welcome to Workflow Engine (#{@company.name})"
```
Comment (hafsabatool7) [inline app/controllers/companies_controller.rb]: Revert to internationalization

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class UsersController < ApplicationController
+  skip_before_action :authenticate_user!
+  def show
+    @company = current_company
+    @users = @company.users.find_by_sequence_num!(params[:id])
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb]: Use cancancan load_and_authorize_resource

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    flash.now[:notif] = "Welcome to Specific User Page of (#{@company.name})"
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb]: No need for this

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  end
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb:11]: Our convention is to include respond_to block even when you are only handling html format

File: app/models/user.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 class User < ApplicationRecord
+
+	sequenceid :company , :users
```
Comment (hafsabatool7) [inline app/models/user.rb]: Indentation

File: app/views/devise/registrations/new.html.erb [hunk 6]
```diff
@@ -1,43 +1,45 @@
-<h2>Sign up</h2>
+<h1 class="h4 text-gray-900 mb-4">Create an Account!</h1>
```
Comment (hafsabatool7) [inline app/views/devise/registrations/new.html.erb]: Use internationalization. Verify throughout the code

File: app/views/devise/registrations/new.html.erb [hunk 7]
```diff
@@ -1,43 +1,45 @@
...
 
 <%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
   <%= render "devise/shared/error_messages", resource: resource %>
 
-  <div class="field">
+  <div class="form-group">
     <%= f.label :name %><br />
-    <%= f.text_field :name%>
+    <%= f.text_field :name, class: "form-control",placeholder: "Enter Name Here", required: true %>
```
Comment (hafsabatool7) [inline app/views/devise/registrations/new.html.erb]: Single quotes and internationalization. Verify everywhere.

File: app/controllers/companies_controller.rb (hunk 1 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/companies_controller.rb]: done

File: app/views/devise/registrations/new.html.erb [hunk 8]
```diff
@@ -18,6 +18,7 @@
     <% if @minimum_password_length %>
       <em>(<%= @minimum_password_length %><%= t('users.character_minimum') %>)</em>
     <% end %><br />
+
```
Comment (hafsabatool7) [inline app/views/devise/registrations/new.html.erb]: Revert

File: db/schema.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
+# This file is auto-generated from the current state of the database. Instead
```
Comment (hafsabatool7) [inline db/schema.rb]: Remove this file from PR

File: config/routes.rb [hunk 10]
```diff
@@ -29,4 +29,5 @@
     post '/user/companies', to: 'home#user_companies', as: 'user_companies'
     root 'home#index'
   end
+  get '/users/:id', to: 'users#show', constraints: { subdomain: /.+/ }
```
Comment (hafsabatool7) [inline config/routes.rb]: resources :user, only: [:show]

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb:11]: done

File: app/models/user.rb (hunk 5 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/user.rb]: done

File: app/views/devise/registrations/new.html.erb (hunk 6 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/new.html.erb]: done

File: app/views/devise/registrations/new.html.erb (hunk 7 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/new.html.erb]: done

File: app/views/devise/registrations/new.html.erb (hunk 8 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/new.html.erb]: done

File: config/routes.rb (hunk 10 above)
Comment (hammad-Ikhlaq-7vals) [reply config/routes.rb]: done

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -1,5 +1,12 @@
 include Pagy::Backend
 class UsersController < ApplicationController
+  load_and_authorize_resource
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb]: Load through sequence number

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (hafsabatool7) [reply app/controllers/users_controller.rb:11]: Do it for index as well

File: app/models/user.rb (hunk 5 above)
Comment (hafsabatool7) [reply app/models/user.rb]: Where?

File: app/models/ability.rb [hunk 12]
```diff
@@ -2,7 +2,7 @@ class Ability
   include CanCan::Ability
 
   def initialize(user)
-
+    @user = user
```
Comment (hafsabatool7) [inline app/models/ability.rb]: Add ability

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -1,5 +1,13 @@
 include Pagy::Backend
 class UsersController < ApplicationController
+  load_and_authorize_resource find_by: :sequence_num
+  def show
+    binding.pry
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb]: Remove binding

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -1,5 +1,13 @@
...
+    respond_to do |format|
+        format.html
+    end
+  end
+
   def index
     @pagy, @users = pagy(User.where(params[:company_id]),  items: 5)
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb]: Respond_to block

File: app/models/ability.rb [hunk 15]
```diff
@@ -2,7 +2,7 @@ class Ability
   include CanCan::Ability
 
   def initialize(user)
-
+    binding.pry
```
Comment (hafsabatool7) [inline app/models/ability.rb]: Remove

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb:11]: done

File: app/models/user.rb (hunk 5 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/user.rb]: Sorry I thought to remove extra line.

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb]: done

File: app/models/ability.rb (hunk 12 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/ability.rb]: done

File: app/controllers/users_controller.rb (hunk 13 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 14 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb]: done

File: app/models/ability.rb (hunk 15 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/ability.rb]: done

File: db/schema.rb (hunk 9 above)
Comment (hammad-Ikhlaq-7vals) [reply db/schema.rb]: done

File: app/controllers/companies_controller.rb [hunk 16]
```diff
@@ -1,7 +1,15 @@
 class CompaniesController < ApplicationController
   # GET '', subdomain: /.+/
   def index
+<<<<<<< HEAD
```
Comment (hafsabatool7) [inline app/controllers/companies_controller.rb]: Resolve this conflict

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -1,7 +1,12 @@
 include Pagy::Backend
 class UsersController < ApplicationController
-  before_action :authenticate_user!
+  load_and_authorize_resource find_by: :sequence_num, through: :current_company
+  def show
+    respond_to { |format| format.html }
+  end
+
   def index
     @pagy, @users = pagy(User.where(params[:company_id]),  items: Company::PAGE_SIZE)
```
Comment (hafsabatool7) [inline app/controllers/users_controller.rb]: User.where(params[:company_id] -> @users (loaded by cancancan)

File: app/models/ability.rb [hunk 18]
```diff
@@ -8,6 +8,17 @@ def initialize(user)
     # *** ALL USERS ***
     return if user.blank?
 
+    if user.account_owner?
+      can :manage, :all
+    elsif user.staff?
+      can :read, Team, id: user.teams.pluck(:id)
+      can :read, Project,  id: user.id
+      can :project_users, Project
```
Comment (hafsabatool7) [inline app/models/ability.rb]: Duplicate code.

File: app/models/ability.rb [hunk 19]
```diff
@@ -8,6 +8,17 @@ def initialize(user)
...
+      can :read, User, sequence_num: user.sequence_num
+    elsif user.admin?
+      can :manage, :all
+      cannot :destroy, User, role: User::ROLES[2]
+    end
```
Comment (hafsabatool7) [inline app/models/ability.rb]: Duplicate code. Adjust your abilities in the relevant block of code.

File: app/models/user.rb [hunk 20]
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+	sequenceid :company ,:users
```
Comment (hafsabatool7) [inline app/models/user.rb]: Indentation

File: config/routes.rb [hunk 21]
```diff
@@ -30,6 +30,7 @@
       end
     end
     get '', to: 'companies#index'
+    devise_for :users
```
Comment (hafsabatool7) [inline config/routes.rb:33]: Why is it required here when it is already present on line 46?

File: app/models/ability.rb [hunk 22]
```diff
@@ -8,6 +8,17 @@ def initialize(user)
     # *** ALL USERS ***
     return if user.blank?
 
+    if user.account_owner?
+      can :manage, :all
+    elsif user.staff?
+      can :read, Team, id: user.teams.pluck(:id)
+      can :read, Project,  id: user.id
+      can :project_users, Project
+      can :read, User, sequence_num: user.sequence_num
```
Comment (hafsabatool7) [inline app/models/ability.rb]: Add the check for companies.

File: app/controllers/companies_controller.rb (hunk 16 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/companies_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 17 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/users_controller.rb]: done

File: config/routes.rb (hunk 21 above)
Comment (hammad-Ikhlaq-7vals) [reply config/routes.rb:33]: resolved

File: app/models/ability.rb (hunk 22 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/ability.rb]: done

File: app/assets/stylesheets/application.scss [hunk 23]
```diff
@@ -65,3 +65,19 @@ $pagination-link-btn-border-color: #ddd;
     }
   }
 }
+
+.profile-head h5{
+    color: #333;
```
Comment (hafsabatool7) [inline app/assets/stylesheets/application.scss]: Store colors in a variable and use that. Like $pagination-link-btn-border-color.

File: app/assets/stylesheets/application.scss [hunk 24]
```diff
@@ -65,3 +65,19 @@ $pagination-link-btn-border-color: #ddd;
...
+}
+
+
+.tab-content profile-tab{
+  margin: 2%;
+}
+
+.col-md-8{
```
Comment (hafsabatool7) [inline app/assets/stylesheets/application.scss]: Do not override default bootstrap classes unless absolute necessary. Use margin classes provided by bootstrap if you want to add margin. Or create your custom class.

File: app/assets/stylesheets/application.scss [hunk 25]
```diff
@@ -65,3 +65,19 @@ $pagination-link-btn-border-color: #ddd;
     }
   }
 }
+
+.profile-head h5{
+    color: #333;
+}
+
+
+.tab-content profile-tab{
```
Comment (hafsabatool7) [inline app/assets/stylesheets/application.scss]: what is profile-tab?

File: app/views/users/show.html.erb [hunk 26]
```diff
@@ -0,0 +1,49 @@
+<div class="col-md-9">
+  <div class="profile-head">
+    <br>
+      <ul class="nav nav-tabs" id="myTab" role="tablist">
```
Comment (hafsabatool7) [inline app/views/users/show.html.erb]: id="myTab" ?
Remove id if not needed. If needed, have descriptive id name and follow underscore convention.

File: app/views/users/show.html.erb [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+          <li class="nav-item">
+              <a class="nav-link active">About</a>
+          </li>
+      </ul>
```
Comment (hafsabatool7) [inline app/views/users/show.html.erb]: Internationalization

File: app/views/users/show.html.erb [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+        <div class="col-md-6">
+            <label><%= t('users.company_label') %></label>
+        </div>
+        <div class="col-md-6">
+            <p><%= @user.company.name %></p>
+        </div>
+    </div>
+    <br>
+    <div>
+      <%= button_to t('shared.update_resource', resource_name: 'User'), edit_user_registration_path, method: :get, class: 'btn btn-info' %>
```
Comment (hafsabatool7) [inline app/views/users/show.html.erb]:
```suggestion
      <%= button_to t('shared.update_resource', resource_name: t('shared.user')), edit_user_registration_path, method: :get, class: 'btn btn-info' %>
```

File: app/assets/stylesheets/application.scss (hunk 23 above)
Comment (hammad-Ikhlaq-7vals) [reply app/assets/stylesheets/application.scss]: Oh sorry, I am not using it anymore. Deleted

File: config/locales/en.yml [hunk 29]
```diff
@@ -129,6 +129,10 @@ en:
     leave_blank_line: "leave blank if you don't want to change it"
     need_for_current_password: 'we need your current password to confirm your changes'
     send_invite: 'Send Invitation'
+    name_label: 'Name:'
+    role_label: 'Role:'
+    email_label: 'Email:'
+    company_label: 'Company:'
```
Comment (hafsabatool7) [inline config/locales/en.yml]: These two are already defined. Please use existing.

File: app/assets/stylesheets/application.scss (hunk 24 above)
Comment (hammad-Ikhlaq-7vals) [reply app/assets/stylesheets/application.scss]: ok

File: app/assets/stylesheets/application.scss (hunk 25 above)
Comment (hammad-Ikhlaq-7vals) [reply app/assets/stylesheets/application.scss]: Removed because not using anymore.

File: app/views/users/show.html.erb (hunk 26 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/users/show.html.erb]: Removed

File: app/views/users/show.html.erb (hunk 27 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/users/show.html.erb]: ok

File: config/locales/en.yml (hunk 29 above)
Comment (hammad-Ikhlaq-7vals) [reply config/locales/en.yml]: done

File: app/views/users/show.html.erb (hunk 28 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/users/show.html.erb]: done

**Pull request merged**